File: backup/Twitch/app/src/main/java/com/exact/twitch/ui/player/HlsPlayerViewModel.java
package com.exact.twitch.ui.player;

import android.app.Application;
import android.net.Uri;
import android.util.Log;

import com.exact.twitch.repository.PlayerRepository;
import com.exact.twitch.ui.OnQualityChangeListener;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.source.hls.HlsManifest;
import com.google.android.exoplayer2.source.hls.HlsMediaSource;
import com.google.android.exoplayer2.source.hls.playlist.HlsMasterPlaylist;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.upstream.DataSource;

import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

abstract class HlsPlayerViewModel extends PlayerViewModel implements OnQualityChangeListener {

    HlsPlayerViewModel(Application context, PlayerRepository playerRepository) {
        super(context, playerRepository);
    }

    Callback<ResponseBody> createPlaylistFetchCallback(DataSource.Factory factory) {
        return new Callback<ResponseBody>() {

            private final String TAG = "PlayerRepository";

            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Log.d(TAG, "Successfully fetched playlist uri");
                //TODO add check if vod is sub only
                HlsMediaSource hlsMediaSource = new HlsMediaSource.Factory(factory).createMediaSource(Uri.parse(response.raw().request().url().toString()));
                preparePlayer(hlsMediaSource);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                Log.e(TAG, "Error getting playlist", t);
            }

        };
    }

    @Override
    public void changeQuality(int index) {
        DefaultTrackSelector.ParametersBuilder parametersBuilder = getTrackSelector().buildUponParameters();
        parametersBuilder.setSelectionOverride(VIDEO_RENDERER, getTrackSelector().getCurrentMappedTrackInfo().getTrackGroups(VIDEO_RENDERER), new DefaultTrackSelector.SelectionOverride(0, index));
        getTrackSelector().setParameters(parametersBuilder);
        setPlayerMode(PlayerMode.NORMAL);
    }

    public void enableAutoMode() {
        DefaultTrackSelector.ParametersBuilder parametersBuilder = getTrackSelector().buildUponParameters();
        parametersBuilder.clearSelectionOverrides();
        getTrackSelector().setParameters(parametersBuilder);
        setSelectedQualityIndex(0);
        setPlayerMode(PlayerMode.NORMAL);
    }

    public void setPlayerMode(PlayerMode playerMode) {
        DefaultTrackSelector.ParametersBuilder parametersBuilder = getTrackSelector().buildUponParameters();
        boolean videoDisabled = false; //NORMAL
        boolean audioDisabled = false;
        switch (playerMode) {
            case AUDIO_ONLY:
                videoDisabled = true;
                audioDisabled = false;
                break;
            case DISABLED:
                videoDisabled = audioDisabled = true;
                break;
        }
        parametersBuilder.setRendererDisabled(VIDEO_RENDERER, videoDisabled);
        parametersBuilder.setRendererDisabled(AUDIO_RENDERER, audioDisabled);
        getTrackSelector().setParameters(parametersBuilder);
    }

    @Override
    public void onTimelineChanged(Timeline timeline, Object manifest, int reason) {
        HlsManifest hlsManifest = (HlsManifest) manifest;
        if (getQualities().getValue() == null) {
            Pattern pattern = Pattern.compile("NAME=\"(.+)\"");
            HlsMasterPlaylist masterPlaylist = hlsManifest.masterPlaylist;
            List<String> tags = masterPlaylist.tags;
            LinkedHashMap<CharSequence, String> urls = new LinkedHashMap<>(tags.size());
            int trackIndex = 0;
            for (int i = 0; i < tags.size(); i++) {
                String tag = tags.get(i);
                Matcher matcher = pattern.matcher(tag);
                if (matcher.find()) {
                    String quality = matcher.group(1);
                    String url = masterPlaylist.variants.get(trackIndex++).url;
                    urls.put(quality.toLowerCase().startsWith("audio") ? "Audio only" : quality, url);
                }
            }
            setUrls(urls);
            LinkedList<CharSequence> list = new LinkedList<>(urls.keySet());
            list.add(list.remove(list.indexOf("Audio only"))); //move audio option to bottom
            getQualities().postValue(list);
        }
    }
}

File: app/src/main/java/com/exact/twitch/ui/player/HlsPlayerViewModel.kt
package com.exact.twitch.ui.player

import android.app.Application
import android.content.Context.MODE_PRIVATE
import com.exact.twitch.ui.Initializable
import com.exact.twitch.ui.OnQualityChangeListener
import com.exact.twitch.ui.player.PlayerMode.AUDIO_ONLY
import com.exact.twitch.ui.player.PlayerMode.DISABLED
import com.exact.twitch.ui.player.PlayerMode.NORMAL
import com.exact.twitch.util.C
import com.google.android.exoplayer2.Timeline
import com.google.android.exoplayer2.source.hls.HlsManifest
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import io.reactivex.disposables.CompositeDisposable
import java.util.LinkedList
import java.util.regex.Pattern

abstract class HlsPlayerViewModel(context: Application) : PlayerViewModel(context, PlayerType.HLS), OnQualityChangeListener {

    companion object {
        const val VIDEO_RENDERER = 0
        const val AUDIO_RENDERER = 1
    }

    val helper = PlayerHelper()

    override fun changeQuality(index: Int, tag: String) {
        when (index) {
            helper.selectedQualityIndex -> return
            in 0 until helper.qualities.value!!.lastIndex -> {
                val parametersBuilder = trackSelector.buildUponParameters()
                when (index) {
                    0 -> parametersBuilder.clearSelectionOverrides() //Auto
                    else -> parametersBuilder.setSelectionOverride(VIDEO_RENDERER, trackSelector.currentMappedTrackInfo?.getTrackGroups(VIDEO_RENDERER), DefaultTrackSelector.SelectionOverride(0, index))
                }
                trackSelector.setParameters(parametersBuilder)
                helper.selectedQualityIndex = index
                changePlayerMode(NORMAL)
                getApplication<Application>().getSharedPreferences(C.USER_PREFS, MODE_PRIVATE).edit().putInt(tag, index).apply()
            }
            helper.qualities.value!!.lastIndex -> changePlayerMode(AUDIO_ONLY)
            else -> changePlayerMode(DISABLED)
        }
    }

    private fun changePlayerMode(playerMode: PlayerMode) {
        val videoDisabled: Boolean
        val audioDisabled: Boolean
        when (playerMode) {
            NORMAL -> {
                videoDisabled = false
                audioDisabled = false
            }
            AUDIO_ONLY -> {
                videoDisabled = true
                audioDisabled = false
            }
            DISABLED -> {
                videoDisabled = true
                audioDisabled = true
            }
        }
        trackSelector.setParameters(trackSelector.buildUponParameters().setRendererDisabled(VIDEO_RENDERER, videoDisabled).setRendererDisabled(AUDIO_RENDERER, audioDisabled))
    }

    override fun onTimelineChanged(timeline: Timeline, manifest: Any?, reason: Int) {
        if (helper.qualities.value == null) {
            val masterPlaylist = (manifest as HlsManifest).masterPlaylist
            val tags = masterPlaylist.tags
            val urls = LinkedHashMap<CharSequence, String>(tags.size)
            val pattern = Pattern.compile("NAME=\"(.+)\"")
            var trackIndex = 0
            tags.forEach {
                val matcher = pattern.matcher(it)
                if (matcher.find()) {
                    val quality = matcher.group(1)
                    val url = masterPlaylist.variants[trackIndex++].url
                    urls[if (!quality.startsWith("audio", true)) quality else "Audio only"] = url
                }
            }
            helper.urls.putAll(urls)
            val list = LinkedList(urls.keys)
            list.add(list.removeAt(list.indexOf("Audio only"))) //move audio option to bottom
            helper.qualities.postValue(list)
        }
    }
}
